### PR TITLE
feat(connector): implement Braintree incoming webhooks in UCS (#17454, #17455)

### PR DESCRIPTION
**Pull request by iemyashasvi**
## What

Implements **IncomingWebhook** support for the **Braintree** connector in UCS (hyperswitch-prism), porting the behaviour 1:1 from **hyperswitch (the Direct gateway), which is the single source of truth**.

Resolves:
- juspay/hyperswitch-cloud#17454 — `EventService.ParseEvent`
- juspay/hyperswitch-cloud#17455 — `EventService.HandleEvent`

## Scope

Single connector, two files: `crates/integrations/connector-integration/src/connectors/braintree.rs` and `braintree/transformers.rs`. No proto / shared-infra changes.

## Behaviour (ported from HS `crates/hyperswitch_connectors/src/connectors/braintree.rs`)

Braintree webhooks are **dispute-only** and **form-urlencoded** (`bt_signature`, `bt_payload`):
- **Signature**: `bt_signature` = `pubkey|sig` pairs joined by `&`; the pair matching the merchant public key is selected. `sig` is verified as `hex(HMAC_SHA1(key = SHA1(private_key), message = bt_payload))`. The webhook secret carries `secret = private_key`, `additional_secret = public_key`.
- **Body**: `bt_payload` base64-decodes to XML, parsed into a dispute `Notification`.
- **Event map** (`get_status`): `dispute_opened→DisputeOpened`, `dispute_lost→DisputeLost`, `dispute_won→DisputeWon`, `dispute_accepted|dispute_auto_accepted→DisputeAccepted`, `dispute_expired→DisputeExpired`, `dispute_disputed→DisputeChallenged`.
- **Object reference**: `PaymentId(ConnectorTransactionId(dispute.transaction.id))` (the parent payment).
- **Dispute content**: amount/currency, `connector_dispute_id`, status, stage (`get_dispute_stage`), reason.
- **Ack**: `"[accepted]"`.

### Primary mode (direct gRPC `EventService`)
- **ParseEvent**: `event_type=WEBHOOK_DISPUTE_OPENED`, reference `dispute{connector_transaction_id="txn_bt_001"}` (no `connector_dispute_id`, matching HS).
- **HandleEvent** (valid signature): `source_verified:true`, `disputes_response{ dispute_status:DISPUTE_OPENED, connector_dispute_id:"dp_bt_001", dispute_stage:ACTIVE_DISPUTE }`, ack `"[accepted]"`.
- **Negatives** (all pass): tampered payload → `source_verified:false`; wrong private key → `source_verified:false`; missing `bt_signature` → graceful `InvalidArgument`, no panic.

### Shadow mode (router runs HS-Direct + shadow-UCS on the same webhook)
`dispute_opened` (xrid `019f1a08-843b-78c1-9c26-9524cbb646bd`):
- `Webhook shadow diff`: `DisputeOpened==DisputeOpened`, `event_type_match=true`, `primary_source_verified == shadow_source_verified` (both `true`).
- validation service `VS_48`: **`keyDiff: {}`**, `typeDiff: {}` — reference matches on both legs (`PaymentId(ConnectorTransactionId(txn_bt_001))`). Only residual is the tautological `content_kind` valueDiff (`direct` vs `unified_connector_service`).

**Primary-through-router** confirmed: flipping the rollout to `execution_mode:primary` and re-firing returned HTTP 200 with `ExecutionPath::UnifiedConnectorService` serving the webhook live.

## Known limitation (pre-existing, not Braintree-specific — out of scope here)

Dispute `amount`/`currency` are computed correctly by the connector on `DisputeWebhookDetailsResponse`, but the **shared** mapper `ForeignTryFrom<DisputeWebhookDetailsResponse> for DisputeResponse` (`crates/types-traits/domain_types/src/types.rs`) hardcodes `dispute_amount: None` for **all** connectors. Surfacing it is a separate shared-infra change, intentionally not bundled here. It does not affect the shadow parity snapshot (which does not compare amount).

## Notes

- No credentials, secrets, signatures, or local config (`config/development.toml`) in the diff. Sample webhook body uses dummy values only.

---

<details>
<summary><b>📋 UCS log proof — incoming webhook received &amp; processed (how it was tested)</b></summary>

**How the webhook is triggered:** Braintree dispute webhooks can't be tested by waiting for Braintree to call `localhost`. The webhook is delivered to the **router's real connector webhook route** `POST /webhooks/{merchant}/{mca}` — the URL Braintree would hit — as a form-urlencoded `bt_signature`/`bt_payload` body whose `transaction.id` is a **real Braintree sandbox payment** (made through the router first), signed with the MCA's keys. The router runs Direct (HS) + shadow-UCS on the same request; the logs below prove UCS received and processed it and agreed with HS.

**Config used to enable shadow / primary** (DB `configs` table; `x-tenant-id: public`, admin api-key):
```
ucs_enabled = true
# shadow:
ucs_rollout_config_braintree_shadow_test_mca_1Yqhdgw6R4EREdAFfec1_Webhooks =
  {"rollout_percent":1.0,"http_url":"http://localhost:8081","https_url":"http://localhost:8081","execution_mode":"shadow"}
# primary (same key, flipped — set via POST /configs/{key} to bust the router config cache):
  {"rollout_percent":1.0,"http_url":"http://localhost:8081","https_url":"http://localhost:8081","execution_mode":"primary"}
```

**1) Router received the webhook at the connector route and dispatched to UCS:**
```text
Webhook gateway decision: gateway=Direct, execution_path=ShadowUnifiedConnectorService
Selected webhook execution path, connector: braintree, execution_path: ShadowUnifiedConnectorService
```
**2) Router shadow comparison (Direct/HS vs shadow-UCS, same webhook):**
```text
Webhook shadow diff, primary_event_type: Some(DisputeOpened), shadow_event_type: Some(DisputeOpened),
  event_type_match: true, primary_source_verified: Some(true), shadow_source_verified: Some(true)
```
**3) UCS (prism grpc-server) RECEIVED the incoming webhook over EventService:**
```text
EventService::handle_event with name: "UCS", flow: "incoming_webhook", service_name: "EventService",
  request_body: {"request_details":{"method":"HTTP_METHOD_POST",
    "uri":"/webhooks/braintree_shadow_test/mca_1Yqhdgw6R4EREdAFfec1",
    "headers":{...}, "body":[...bt_signature=<pubkey>|<redacted>&bt_payload=<base64 dispute XML, txn id=txn_***>...]}}
```
**4) UCS PROCESSED it — resolved event type + verified source (response):**
```text
ParseEvent  -> ..."resource":{"Dispute":{"connector_transaction_id":"txn_***"}}},"event_type":"WEBHOOK_DISPUTE_OPENED"}
HandleEvent -> ..."disputes_response":{...,"dispute_status":"DISPUTE_OPENED",...,"dispute_message":"fraud"}},"source_verified":true
```
(Validation service VS_48: `bodyComparison.keyDiff: {}` — UCS outcome matches HS Direct exactly; only the benign `content_kind` provenance label differs.)

</details>

---

<details>
<summary><b>📡 Real-key webhook E2E — Braintree dispute → ngrok → router → UCS (shadow + primary)</b></summary>

**How / honest caveat:** Braintree dispute webhooks can only be *delivered by Braintree's servers* if a destination URL is configured in the **gateway control panel** and a real dispute is triggered — neither is API-scriptable (only API keys are available, not control-panel login). So the strongest achievable test: a dispute notification **signed with the REAL Braintree merchant private key** (byte-identical to Braintree's SDK `sampleNotification` signing), referencing a **REAL Braintree sandbox payment** (made through the router), delivered through the **same ngrok tunnel** → router.

**ngrok request inspector:**
```text
POST https://<ngrok>/webhooks/<merchant>/<mca>  ->  200   [accepted]
```

**Shadow** (rollout `execution_mode=shadow`):
```text
router: Webhook gateway decision: gateway=Direct, execution_path=ShadowUnifiedConnectorService
router: Webhook shadow diff: primary_event_type=Some(DisputeOpened), shadow_event_type=Some(DisputeOpened),
        event_type_match=true, primary_source_verified=Some(true), shadow_source_verified=Some(true)
UCS:    EventService::handle_event  flow="incoming_webhook"  uri="/webhooks/<merchant>/<mca>"
UCS:    response -> "event_type":"WEBHOOK_DISPUTE_OPENED"   "source_verified":true
```
Both HS-Direct **and** shadow-UCS verified the **real-Braintree-key signature** against the MCA's webhook keys, resolving the dispute to the real sandbox transaction.

**Primary** (rollout `execution_mode=primary`, re-delivered):
```text
router: Webhook gateway decision: gateway=UnifiedConnectorService, execution_path=UnifiedConnectorService   (HTTP 200 [accepted])
```

_Rollout restored to shadow afterward. No api keys / private keys / signatures / real ids shown (sandbox transaction id masked; ngrok host generalized)._

</details>

### CODE REVIEW

**Comment by XyneSpaces** [comment]
## 🔒 Security Review: Braintree Webhook Implementation

I've analyzed this PR implementing Braintree incoming webhooks. Overall, the implementation follows the established patterns for webhook verification with HMAC-SHA256 signature validation. Here are my findings:

### ✅ Correctly Implemented
- Uses `Secret<String>` for cryptogram storage (PII protection)
- Proper error propagation with `error_stack`
- NTID requirement enforced for MIT flows
- Card type mapping uses safe `try_from` conversion

### ⚠️ Suggestions
1. **Constant-time comparison**: Ensure signature verification uses constant-time comparison to prevent timing attacks
2. **Test coverage**: Add test cases for webhook signature verification edge cases (expired signatures, replay attacks)
3. **Documentation**: Consider adding a brief comment explaining the Braintree signature format for future maintainers

### 🎯 Recommendation
Ready to merge after addressing the constant-time comparison concern if not already present in the crypto utilities.

**Comment by XyneSpaces** [comment]
:bulb: Webhook signature verification uses `crypto::SignMessage` — ensure the implementation uses constant-time comparison to prevent timing attacks on the signature verification.

Verify that `crypto::verify_signature` or equivalent uses a constant-time comparison primitive rather than standard string/array equality.

**Comment by XyneSpaces** [comment]
:rotating_light: **Critical: Missing webhook signature verification function**

The `verify_webhook_source` method in `braintree.rs` calls `braintree::verify_braintree_signature(...)` at line 116, but this function is **not implemented** in the PR. The `transformers.rs` file only shows the webhook data types (`BraintreeWebhookBody`, `BraintreeDispute`, etc.) but lacks the actual signature verification logic.

**Security Impact**: Without signature verification, any unauthenticated caller can trigger dispute state transitions by sending crafted webhook payloads to the endpoint.

**Required Fix**: Implement `verify_braintree_signature` in `transformers.rs` using constant-time comparison (e.g., `subtle::ConstantTimeEq`) to prevent timing attacks:

```rust
fn verify_braintree_signature(
    signature: &str,
    payload: &str,
    public_key: &str,
    private_key: &str,
) -> CustomResult<bool, UnifiedError> {
    // Parse signature, compute HMAC-SHA1, constant-time compare
}
```

Reference the Stripe connector's `verify_webhook_source` implementation for the pattern.

**Review by deepanshu-iiitu — DISMISSED** [review]

**Comment by XyneSpaces** [comment]
This is a test comment to check the issue comments structure.

File: crates/integrations/connector-integration/src/connectors/braintree/transformers.rs [hunk 1]
```diff
@@ -2741,6 +2743,149 @@ pub struct DisputeEvidence {
...
+        .change_context(domain_types::errors::WebhookError::WebhookBodyDecodingFailed)
+}
+
+// Base64-decodes the (newline-stripped) `bt_payload` and parses the XML `Notification`.
+pub(crate) fn decode_webhook_payload(
+    payload: &[u8],
+) -> Result<Notification, Report<domain_types::errors::WebhookError>> {
+    let decoded_response = super::BASE64_ENGINE
+        .decode(payload)
+        .change_context(domain_types::errors::WebhookError::WebhookBodyDecodingFailed)?;
```
Comment (AmitsinghTanwar007) [inline crates/integrations/connector-integration/src/connectors/braintree/transformers.rs]: try to use attach printable with change context so that while debugging things are clear

File: crates/integrations/connector-integration/src/connectors/braintree/transformers.rs [hunk 2]
```diff
@@ -2741,6 +2743,149 @@ pub struct DisputeEvidence {
     pub url: url::Url,
 }
 
+// Maps the Braintree notification `kind` to the prism webhook event type.
+// Ports HS `get_status` (hyperswitch braintree/transformers.rs `get_status`) 1:1.
+pub(crate) fn get_status(status: &str) -> connector_types::EventType {
```
Comment (AmitsinghTanwar007) [inline crates/integrations/connector-integration/src/connectors/braintree/transformers.rs]: why this needs a pub(crate)

File: crates/integrations/connector-integration/src/connectors/braintree/transformers.rs (hunk 1 above)
Comment (iemyashasvi) [reply crates/integrations/connector-integration/src/connectors/braintree/transformers.rs]: Done — added `.attach_printable(...)` to every webhook decode step so failures carry a clear message: url-decoding the `bt_signature`/`bt_payload` envelope, base64-decoding `bt_payload`, the UTF-8 conversion, and the `Notification` XML parse.

File: crates/integrations/connector-integration/src/connectors/braintree/transformers.rs (hunk 2 above)
Comment (iemyashasvi) [reply crates/integrations/connector-integration/src/connectors/braintree/transformers.rs]: Good catch — narrowed to `pub(super)`. These helpers are only called from the parent `braintree.rs` (`IncomingWebhook` impl), never elsewhere in the crate, so `pub(super)` is the minimal visibility they need. Applied the same to the other braintree webhook helpers for consistency.

**Review by AmitsinghTanwar007 — DISMISSED** [review]

**Review by AmitsinghTanwar007 — APPROVED** [review]

**Review by deepanshu-iiitu — APPROVED** [review]
